deposit-form: make use of `Overridable` consistent

palkerecsenyi opened the pull request:
Closes https://github.com/inveniosoftware/invenio-app-rdm/issues/3103

**Please release at the same time as https://github.com/inveniosoftware/react-invenio-forms/pull/296 and https://github.com/inveniosoftware/react-invenio-forms/pull/298**

---

### Description

* Previously, the fields in the deposit form had inconsistent use of `Overridable` making it difficult to understand how to override them or their attributes in a predictable manner.

* This PR aims to make the use of `Overridable` more uniform across the fields, allowing granular overriding of fields and their subcomponents.

* It also establishes a [set of common props](https://github.com/palkerecsenyi/invenio-rdm-records/blob/bd1a0659accbbf651f7104f582bb2799bfc31473/invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/fields/common/fieldComponents.js#L12-L28) that are used by (nearly) all fields and so anyone implementing an override can expect them to function consistently when applied to any field. The common props can be overridden for any field component, in addition to the component-specific props.

* The `hidden` prop is handled automatically by the new [`createCommonFieldDepositComponent` function](https://github.com/palkerecsenyi/invenio-rdm-records/blob/bd1a0659accbbf651f7104f582bb2799bfc31473/invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/fields/common/fieldComponents.js#L30-L42) through which all field components are now exported.
  * When a field is set to `hidden`, its value is _not_ submitted to the API upon form submission. **However**, when a record already has a given field which is then later set to hidden, that value will remain in the database unchanged after submitting the form.
  * This is because Formik **always submits all the values** provided in [`initialValues`](https://github.com/inveniosoftware/invenio-rdm-records/blob/3b7c4b6ec55742cd371bde01a487096ae9923dff/invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/api/DepositBootstrap.js#L133) to the server, even if an input controlling a value is never rendered in the form.

* The `required` prop also does not affect the data model currently. It simply affects the view, hiding/showing the '*' symbol. For example, setting the `LicenseField` to required does not mean the user is prevented from submitting the form without specifying the field, but the asterisk is indeed shown:

  ```javascript
  // my_rdm_instance/assets/js/invenio_app_rdm/overridableRegistry/mapping.js

  import { parametrize } from "react-overridable"
  import { LicenseField } from "@js/invenio_rdm_records/src/deposit/fields"
  
  export const overriddenComponents = {
    "InvenioRdmRecords.DepositForm.LicenseField": parametrize(LicenseField, {
      label: "What is the license?",
      required: true,
    }),
  }
  ```

  <img width="457" height="116" alt="image" src="https://github.com/user-attachments/assets/d5f4a010-133c-45b8-834b-522ad96ff75b" />

  We need to assess whether this should be the intended behaviour or if we should implement some sort of additional change on the backend to ensure validation. Formik currently does not perform validation beyond some very basic things.

### Next steps

See the parent issue https://github.com/inveniosoftware/invenio-app-rdm/issues/3100

utnapischtim commented on .gitignore:
```diff
@@ -77,3 +77,9 @@ pip-wheel-metadata/
 node_modules
+
+# Needed for certain vim configs to work properly
+tsconfig.json
```
nit: i am unsure if we should add this to .gitignore.
there is an option to create a user `~/.gitignore` file and add personal configuration files there.

palkerecsenyi replied:
Thanks for raising this, I was unsure what the gitignore convention was as there are some other editor-specific ignores in there, but I'm happy to remove this of course.

utnapischtim replied:
to be honest, my feeling about the .gitignore additions is based on the experience that it has not been changed that often in the past years. so the editor-specific ignores are there since a long time.

sakshamarora1 commented on invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/fields/CopyrightsField/CopyrightsField.js:
```diff
@@ -7,40 +7,69 @@
-        label={
-          <FieldLabel htmlFor={fieldPath} icon="copyright outline" label={label} />
-        }
+      <Overridable
```
Why do we need this when we already have `createCommonDepositFieldComponent` overridable definition?

sakshamarora1 commented on invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/fields/DatesField/DatesField.js:
```diff
@@ -153,17 +176,25 @@ DatesField.propTypes = {
 };
+
+export const DatesField = createCommonDepositFieldComponent(
+  "InvenioRdmRecords.DatesField",
```
```suggestion
  "InvenioRdmRecords.DepositForm..DatesField",
```

sakshamarora1 commented on invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/fields/DescriptionsField/components/AdditionalDescriptionsField.js:
```diff
@@ -15,79 +15,116 @@ import { emptyAdditionalDescription } from "./initialValues";
+                      </Overridable>
+                    </Grid.Column>
+                    <Grid.Column mobile={16} tablet={6} computer={4}>
+                      <Overridable id="InvenioRdmRecords.DescriptionsField.Additional.remove">
```
Do we really need this?

sakshamarora1 commented on invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/fields/DescriptionsField/components/AdditionalDescriptionsField.js:
```diff
@@ -15,79 +15,116 @@ import { emptyAdditionalDescription } from "./initialValues";
+                            onClick={() => arrayHelpers.remove(indexPath)}
+                          >
+                            <Overridable id="InvenioRdmRecords.DescriptionsField.Additional.remove.icon">
+                              <Icon name="close" />
```
I see the usecase, but why not just give a className here and override it using css?

kpsherva replied:
agreed, I think we are going too granular with the overridables here

sakshamarora1 commented on invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/fields/RelatedWorksField/RelatedWorksField.js:
```diff
@@ -13,113 +13,178 @@ import {
-                  required
-                  width={2}
-                />
+                  <Overridable
```
Wouldn't it just make sense to pass an overridable id to all these types of fields and define a dynamic overridable tag inside the SelectField and it would reduce the amount of code.

utnapischtim commented on invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/fields/CreatibutorsField/CreatibutorsField.js:
```diff
@@ -11,14 +11,20 @@ import PropTypes from "prop-types";
+  FeedbackLabel,
+  FieldLabel,
+  createCommonDepositFieldComponent,
+  mandatoryFieldCommonProps,
```
may i ask where `mandatoryFieldCommonProps` is defined in `react-invenio-forms`. i can't find it.

utnapischtim replied:
ahhh, thanks, i thought i checked all PR's from you in react-invenio-forms. i missed that one.

palkerecsenyi commented:
JS test is currently failing due to a dependency on new functions in https://github.com/inveniosoftware/react-invenio-forms/pull/298 which are not yet in the public NPM package.

zzacharo commented on invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/fields/License/LicenseModal.js:
```diff
@@ -222,21 +221,29 @@ export class LicenseModal extends Component {
-                  <TextAreaField
-                    fieldPath="selectedLicense.description"
+                  <Overridable
+                    id="InvenioRdmRecords.DepositForm.LicenseField.Modal.description"
```
We can remove the overridable ids for these internal ones as we added one for the top level field.

zzacharo commented on invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/fields/License/LicenseModal.js:
```diff
@@ -222,21 +221,29 @@ export class LicenseModal extends Component {
+                    />
+                  </Overridable>
+                  <Overridable
+                    id="InvenioRdmRecords.DepositForm.LicenseField.Modal.link"
```
ditto

zzacharo commented on invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/fields/TitlesField/AdditionalTitlesField.js:
```diff
@@ -9,85 +9,98 @@
     return (
-      <ArrayField
-        addButtonLabel={i18next.t("Add titles")}
+      <Overridable
```
Let's remove this as it is part of the TitlesField overridable component.

zzacharo commented on invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/fields/TitlesField/TitlesField.js:
```diff
@@ -8,37 +8,58 @@
+            className="title-field"
+            optimized
+          />
+          <Overridable
```
Let's remove also this.

zzacharo commented on invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/fields/TitlesField/TitlesField.js:
```diff
@@ -8,37 +8,52 @@
-      </>
+      <Overridable
+        id="InvenioRdmRecords.DepositForm.TitlesField.Container"
+        optimized
```
I saw you removed `optimized` in another example from the overridable props. Maybe do it also here and if it is applicable to other places?

palkerecsenyi replied:
`optimized` is a prop that reasonably should be overridable. It controls whether [`FastField`](https://formik.org/docs/api/fastfield) instead of [`Field`](https://formik.org/docs/api/field) is rendered, which improves performance on large forms such as our deposit form but can also cause a lot of bugs, so implementors can definitely have some use cases where they want to disable this. I will look at implementing support for `optimized` overrides across the deposit form fields so we do it more consistently.

palkerecsenyi commented on invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/fields/RelatedWorksField/RelatedWorksField.js:
```diff
@@ -80,46 +98,59 @@ export class RelatedWorksField extends Component {
-                  fieldPath={`${fieldPathPrefix}.resource_type`}
-                  labelIcon="" // Otherwise breaks alignment
                   options={options.resource_type}
+                  labelIcon=""
```
Keeping the `labelIcon` was a mistake, I removed it from `<ResourceTypeField>` after testing it and finding it made no difference anymore. I meant to also remove it from the overridable, so I will do this now.

palkerecsenyi committed:
feat(deposit)!: consistently apply Overridable to all form fields

* Previously, the fields in the deposit form had inconsistent use of
`Overridable` making it difficult to understand how to override them or
their attributes in a predictable manner.

* This PR aims to make the use of Overridable more uniform across the
fields, allowing more granular (but not too granular) overriding of
fields and their subcomponents. It uses a newly created common set of
props for all form fields, improving the developer experience by making
parameterization patterns more predictable. See
https://github.com/inveniosoftware/react-invenio-forms/pull/298/files#diff-4e411d0626db11a2e9708ca5291378bc0162fce01c2238531e654c958f9256e4
for details.

* These common props and overridables have also been implemented for the
contrib custom fields Imprint, Journal, Thesis, and Meeting, using a
dynamic overridable ID passed down as a prop from Python, in compliance
with the methods used in react-invenio-forms.

* The `hidden` prop is handled automatically by the new
showHideOverridable function through which all field components are now
exported, also defined in react-invenio-forms as above. When a field is
set to hidden, its value is not submitted to the API upon form
submission. However, when a record already has a given field which is
then later set to hidden, that value will remain in the database
unchanged after submitting the form.

* The required prop also does not affect the data model. It simply
affects the view, hiding/showing the '*' symbol. For example, setting
the LicenseField to required does not mean the user is prevented from
submitting the form without specifying the field, but the asterisk is
indeed shown.

* The versions of react-invenio-forms and react-overridable have been
bumped to incorporate the new common props objects and the new devtools.

BREAKING CHANGE: a number of existing Overridable IDs have been renamed
to improve compliance with naming conventions and further improve the
developer experience. The full list of renamed IDs is in CHANGES.rst. We
originally discussed removing some too-granular Overridables in the
InvenioRDM telecon of 2025-09-09, but in the end we decided *not* to
remove any.

palkerecsenyi committed:
release: v20.2.0
